src/amr/solvers/solver_ppc_model_view.hpp

```diff
@@ -0,0 +1,293 @@
+        {
+            auto _ = core::SetLayout(layouts[i], ampere_);
+            ampere_(*B[i], *J[i]);
+        }
+    }
+    core_type ampere_;
+};
+
+template<typename GridLayout>
+class OhmTransformer
+{
+    using core_type = PHARE::core::Ohm<GridLayout>;
+
+public:
+    explicit OhmTransformer(initializer::PHAREDict const& dict)
+        : ohm_{dict}
+    {
+    }
+
+    template<typename GridLayouts, typename VecFields, typename Fields>
+    void operator()(GridLayouts const& layouts, Fields const& n, VecFields const& Ve,
+                    Fields const& Pe, VecFields const& B, VecFields const& J, VecFields& Enew)
+    {
+        assert_equal_sizes(n, Ve, Pe, B, J, Enew);
+        for (std::size_t i = 0; i < B.size(); ++i)
+        {
+            auto _ = core::SetLayout(layouts[i], ohm_);
+            ohm_(*n[i], *Ve[i], *Pe[i], *B[i], *J[i], *Enew[i]);
+        }
+    }
+
+    core_type ohm_;
+};
+
+
+
+template<typename HybridModel_>
+class HybridPPCModelView : public ISolverModelView
+{
+public:
+    using This             = HybridPPCModelView<HybridModel_>;
+    using HybridModel_t    = HybridModel_;
+    using HybridModel_args = typename HybridModel_t::type_list::Tuple;
+    using IPhysicalModel_t = typename HybridModel_t::Interface;
+    using patch_t          = typename HybridModel_t::patch_t;
+    using Electrons        = typename HybridModel_t::electrons_t;
+    using level_t          = typename HybridModel_t::amr_types::level_t;
+    using Electromag       = typename HybridModel_t::electromag_type;
+    using Ions             = typename HybridModel_t::ions_type;
+    using ParticleArray_t  = typename Ions::particle_array_type;
+    using Particle_t       = typename ParticleArray_t::value_type;
+    using VecFieldT        = typename HybridModel_t::vecfield_type;
+    using FieldT           = typename HybridModel_t::field_type;
+    using GridLayout       = typename HybridModel_t::gridlayout_type;
+    using Faraday_t        = FaradayTransformer<GridLayout>;
+    using Ampere_t         = AmpereTransformer<GridLayout>;
+    using Ohm_t            = OhmTransformer<GridLayout>;
+
+    struct PatchState_t;
+
+    using value_type = PatchState_t;
+
+    template<bool isConst = false>
+    struct iterator;
+
+    HybridPPCModelView(level_t& level, IPhysicalModel_t& model)
+        : model_{dynamic_cast<HybridModel_t&>(model)}
+    {
+        onRegrid(level, model_);
+    }
+
+    void onRegrid(level_t& level, HybridModel_t& hybridModel);
+
+    auto begin() { return iterator</*const=*/false>{*this}; }
+    auto begin() const { return iterator</*const=*/true>{*this}; }
+
+    auto end() { return iterator</*const=*/false>{*this, states.size()}; }
+    auto end() const { return iterator</*const=*/true>{*this, states.size()}; }
+
+    auto& model() { return model_; }
+    auto& model() const { return model_; }
+
+    HybridModel_t& model_;
+    std::vector<core::aggregate_adapter<PatchState_t>> states;
+
+
```

Comment:
```cpp 
/*these vectors allow to access patch data per quantity
   they are kept in synced with "states" so that client code (e.g. Faraday etc.)
   can still have signatures revealing which quantities they use, and not take the states vector entirely */
```